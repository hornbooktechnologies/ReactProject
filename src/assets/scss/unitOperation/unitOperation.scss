@import "../_color";
@import "../_variables";

.bg-grey-blue-40-a10 {
  background-color: rgba($grey_blue_40, 0.1);
}
.color-grey-blue-60 {
  color: $grey_blue_60;
}
.min-w-1190 {
  min-width: 1190px;
}
.m-w-110 {
  min-width: 110px;
}
.max-height-500 {
  max-height: $val10 * 50;
}
.font-weight-500 {
  font-weight: 500;
}
// uk Table start
.uk-table {
  &.uk-table-divider {
    th,
    td {
      border: 1px $solid $grey_blue_40;
      text-align: left;
      padding: $val10 $val10 - 2;
      color: $white;
      font-size: $val15 - 1;
      .icn {
        position: $relative;
        top: 2px;
        &.icn-right {
          float: $right;
        }
      }
    }
    th {
      background-color: rgba($white, 0.1);
      font-size: $val15 + 1;
      vertical-align: $middle;
      text-transform: $inherit;
      font-weight: 500;
      text-align: $center;
    }
  }
  &.uk-text-center {
    th,
    td {
      text-align: $center;
    }
  }
  &.uk-text-right {
    th,
    td {
      text-align: $right;
    }
  }
  &.uk-table-sm {
    th {
      font-size: $val15 - 2 !important;
    }
    td {
      font-size: $val15 - 3 !important;
    }
  }
  &.uk-table-middle {
    th,
    td {
      vertical-align: $middle !important;
    }
  }
}
.min-w-86 {
  min-width: 86px;
  width: fit-content;
  margin-top: $val0;
}

@media (max-width: 767px) {
  .media-mt-30 {
    margin-top: 3 * $val10;
  }
}

