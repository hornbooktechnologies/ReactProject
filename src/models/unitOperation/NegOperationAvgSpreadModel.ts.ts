export type NegOperationAvgSpreadModel = {negOperationAvrgSpread:AverageSpreadCardModel[]}
export type NegativeOperationAvgSpreadModel = AverageSpreadCardModel[]

export type AverageSpreadCardModel = {
  title: string,
  Suffix: string,
  plan: AverageSpreadRecordModel
  actual_forcast?: AverageSpreadRecordModel
  yearStartToPresent?: AverageSpreadRecordModel
}

export type AverageSpreadRecordModel = {
  header?:string | undefined
  previousValue?: number | undefined | string
  suffix?: string 
  value?: number | undefined | string
}
