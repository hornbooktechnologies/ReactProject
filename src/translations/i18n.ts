import { initReactI18next } from 'react-i18next'
import i18n from 'i18next'
import LanguageDetector from 'i18next-browser-languagedetector'
import { TRANSLATIONS_EN } from './en/Translations'
import { TRANSLATIONS_JA } from './ja/Translations'

i18n
  .use(LanguageDetector)
  .use(initReactI18next)
  .init({
    lng: 'en',
    resources: {
      en: {
        translation: TRANSLATIONS_EN,
      },
      ja: {
        translation: TRANSLATIONS_JA,
      },
    },
  })

export default i18n
