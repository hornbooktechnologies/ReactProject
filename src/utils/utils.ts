import { i18n } from "i18next";
import { DateTime } from "luxon";
import moment from "moment";
import { DonutChartCustomData, DonutChartDataModel } from "../models/DonutChartDataModel";
import { DropDownData } from "../models/DropdownModel";
import { CommonApiMap } from "./apiUtils/ApiParamMap";
import { AppConstant, DateformatConstant, MathConstant, NumberConstant, OutageOutputConstant, TIMEZONE } from "./AppConstants";
import { cssClassName } from "./CssConstants";
import { localStorageUtils } from "./LocalStorageUtils";

export const GetDonutChartSeriesAndValueForCurrentStatus = (i18n: i18n, customData: DonutChartCustomData, seriesData: any): DonutChartDataModel => {
    return {
        ...customData,
        series1: {
            title: i18n.t("LABELS.PLAN"),
            seriesValue: seriesData?.Plan || 0,
            prefix: seriesData?.Prefix || '',
            suffix1: seriesData?.Suffix || '',
            suffix2: seriesData?.Suffix2 || '',
        },
        series2: {
            title: i18n.t("LABELS.FORECAST"),
            seriesValue: seriesData?.Forecast || 0,
            prefix: seriesData?.Prefix || '',
            suffix1: seriesData?.Suffix || '',
            suffix2: seriesData?.Suffix2 || '',
        },
        value: getDonutChartValue(seriesData),
        plan: seriesData?.Plan,
        forecast: seriesData?.Forecast,
    } as DonutChartDataModel;
}

function getDonutChartValue(seriesData: any) {
    return (seriesData?.Forecast / seriesData?.Plan) * 100;
}

export const getDateTimeEpoch = (datetime: string): number => {
  return new Date(datetime).getTime() + TIMEZONE.JPN_TIMEZONE
}

export const getFiscalYear = (): number => {
  if (new Date().getMonth() + 1 >= 1 && new Date().getMonth() + 1 <= 3) {
    return new Date().getFullYear() - 1
  } else {
    return new Date().getFullYear()
  }
}

export const getFiscalYearTimestamp = (fiscalYear: number): number => {
  return new Date(`${fiscalYear}${DateformatConstant.FISCAL_YEAR}`).getTime();
}

export const groupBy = (array, key) => {
  return array.reduce((result, currentValue) => {
    (result[currentValue[key]] = result[currentValue[key]] || []).push(
      currentValue
    );
    return result;
  }, {});
};

export const getGeneratorDropdownData = (generatorList: any[]) => {
  return generatorList?.map((generator) => {
    return { 
      label: generator[CommonApiMap.generatorDisplayName],
      value: getFormattedUnitId(generator[CommonApiMap.powerPlantUnitId])
    } as DropDownData
  })
}

export const getFormattedUnitId = (unitId: string) => {
  return unitId?.slice(0,3);
}

export const formatNumberToFractions = (value: number|string |undefined|null, noOfDecimals: number) => {
  let noOfDecimalsVal = noOfDecimals ? noOfDecimals : 1
  if (!value || value === "undefined") {value = MathConstant.DEFAULT_FRACTION_STRING;}
  return parseFloat(value.toString())?.toFixed(noOfDecimalsVal);
}

export const getCutOffUnixEpochSeconds = (): number => {
  const unixTime = new Date().getTime();
  return Math.floor(unixTime / MathConstant._1000);
}

export const getCutoffFractionTime = (): number => {
  const unixTime = new Date().getTime();
  return Math.trunc(unixTime / MathConstant._1000);
}

export const getOtherUnits = (selectedUnit: string | null) => {
  const allUnits = (localStorageUtils.getSelectedPlantUserPreferance())?.generators;
  return allUnits?.filter(unit => getFormattedUnitId(unit[CommonApiMap.powerPlantUnitId]) !== selectedUnit);
}

export const getUnitNameById = (unitId: string) => {
  const allUnits = (localStorageUtils.getSelectedPlantUserPreferance())?.generators;
  const unit = allUnits && allUnits.find(x => x[CommonApiMap.powerPlantUnitId] === `${unitId}0`);
  return unit && unit[CommonApiMap.generatorDisplayName] || '' || undefined;
}

export const getCutoffTime = () => {
  return DateTime.fromMillis(getCutOffUnixEpochSeconds() * MathConstant._1000, {
    zone: TIMEZONE.ASIA_TOKYO,
  }).toFormat(DateformatConstant.CUTOFF_FORMAT);
}

export const formatDate = (date: string) => {
  return moment(date).format(DateformatConstant.YYYY_M_D_HH_MM);
}

export const getSelectedUnitID= () => {
  const selectedUserPreference = localStorageUtils.getSelectedPlantUserPreferance()
  const units = getGeneratorDropdownData(selectedUserPreference?.generators)
  const selectedUnitId = localStorageUtils.getSelectedUnitId() || getFormattedUnitId(units && units[0].value)
  return selectedUnitId;
}

export const generatorDropdownUnit =()=>{
  const selectedUserPreference = localStorageUtils.getSelectedPlantUserPreferance()
  return getGeneratorDropdownData(selectedUserPreference?.generators)
}

export const calculateDiff = (firstVal: number | string,secondVal:number | string) => {
  let diff  
  if (firstVal && secondVal) {
    diff = firstVal - secondVal
  }
  return diff.toFixed(NumberConstant.TWO)
}

export const prefixSign = (value: number|string,variation: string) => {
  let prefixSign  
 
    if(value === 0 || value < 0 || value === null) {
      prefixSign = ''
    }
    else if (value > 0 || variation === OutageOutputConstant.POSITIVE_VAL) {
      prefixSign = AppConstant.PLUS_SYMBOL
    }
    else if (variation === AppConstant.MINUS_SYMBOL) {
      prefixSign = AppConstant.MINUS_SYMBOL
    }
  
  return prefixSign 
}

export const IsappearTooltip = (value:boolean) => {
  let show  
  if(value === true){
    show = cssClassName.CUSTOM_TOOLTIP
  }
  else {
    show = ''
  }
  return show 
}

export const IstextHighlight = (value:boolean) => {
  let show  
  if(value === true){
    show = cssClassName.TEXT_HIGHLIGHT
  }
  else {
    show = ''
  }
  return show 
}

export const addCustomClass = (value:boolean,className:string|undefined) => {
  let customClass  
  if(value === true){
    customClass = className + cssClassName.TOOLTIP_CUSTOM
  }
  else {
    customClass = ''
  }
  return customClass 
}

export const setTooltipPosition = (className:string|undefined) => {
  let addClass  
  if(className !== cssClassName.TOOLTIP_BOTTOM){
    addClass = cssClassName.TOOLTIP_TOP
  }
  else {
    addClass = ''
  }
  return addClass 
}

export const getDiffOfValues = (ValueA:number|string|null,ValueB:number|string|null) => {
  let diffOfValues 
  if( ValueA && ValueB ) {
    diffOfValues = ValueA !== ValueB
  }
   return diffOfValues
}