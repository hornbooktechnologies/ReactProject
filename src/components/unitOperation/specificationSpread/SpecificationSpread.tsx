import React, { useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { ApiParamUtils } from '../../../utils/apiUtils/ApiParamUtils'
import useStateRef from 'react-usestateref'
import {
  AnnualTargetConstant,
  AppConstant,
  DateformatConstant,
  ChartConstant,
  UserPrefModelConstant,
} from '../../../utils/AppConstants'
import {
  generatorDropdownUnit,
  getFormattedUnitId,
  getGeneratorDropdownData,
} from '../../../utils/utils'
import LinechartDropDown from '../../common/Dropdown'
import GeneratorDropDown from '../../common/GeneratorDropDown'
import DataChangeOptions from '../../common/DataChangeOptions'
import { FetchData } from './LinechartData'
import Moment from 'moment'
import SpreadLineChart from './SpreadLineChart'
import { localStorageUtils } from '../../../utils/LocalStorageUtils'
import {
  SpecificationSpreadDataModel,
  SpreadLineChartBaseDataModel,
  titleDropdownModel,
} from '../../../models/unitOperation/UnitOperationModel'
import { getCutoffFractionTime } from '../../../utils/utils'
import { linechartDataModel } from '../../../models/LinechartModel'

const SpecificationSpread = (props: SpecificationSpreadDataModel) => {
  const { i18n } = useTranslation()
  const selectedUserPreference = localStorageUtils.getSelectedPlantUserPreferance()
  const units = getGeneratorDropdownData(selectedUserPreference?.generators)
  const [selectedUnitId, setSelectedUnitId] = useState(
    localStorageUtils.getSelectedUnitId() ||
      getFormattedUnitId(units && units[0].value)
  )
  const [selectValue, setSelectValue] = useState(
    AnnualTargetConstant.ANNUAL_DATA
  )
  const [startDate, setStartDate] = useState(new Date(props.time * 1000))
  const [linechartData, setLinechartData] = useState<
    linechartDataModel | undefined
  >()
  const [titleDropdown, setTitleDropdown] = useState([])
  const [selectedTitleDropdown, setSelectedTitleDropdown] = useState('')
  const [ind, setIndex, indexref] = useStateRef<any>()
  const powerPlantUnitId =  UserPrefModelConstant.POWER_PLANT_UNIT_ID
  const powerTitle = UserPrefModelConstant.POWER_TITLE
  let propsData  = props.spreadOperation

  useEffect(() => {
    if (linechartData && linechartData.graphData === null && ind === 0) {
      props.noDataHandler(getCutoffFractionTime(), selectValue)
    } else {
    }
  }, [linechartData, props, selectValue, startDate, ind])

  let spreadLineChartdata: SpreadLineChartBaseDataModel[]  = []
  let dataEstList = []
  let list = []
  var SpreadList

  if (linechartData && linechartData.graphData) {
    let parsed = JSON.parse(linechartData.graphData)
    let linechartDataObj = {
      Period: String,
      Spread: Number,
      Estimates: Number,
      TimeStamp: Number,
      powerPlantUnitId: String,
      powerTitle: String,
    }
  for (const key in parsed.Period) {
      let graphData = Object.create(linechartDataObj)
      if (parsed.Period[key]) {
        graphData.Period = parsed.Period[key]
        graphData.Spread = parsed.Spread[key]
        graphData.powerPlantUnitId = parsed[powerPlantUnitId]
        graphData.poweTitle = parsed[powerTitle]
        let epoch = new Date(parsed.Period[key])
        graphData.TimeStamp = epoch.getTime()
        if (selectValue === ChartConstant.MONTHLY_DATA) {
          if (Moment(graphData.Period).format(DateformatConstant.DDD) === DateformatConstant.SUN) {
            list.push(
              Moment(graphData.Period).format(DateformatConstant.YYYY_MM_DDT00)
              )
            }
          }
          spreadLineChartdata.push(graphData)
      }
    }

    for (const d in parsed.Estimates) {
      let dataEs = Object.create(linechartDataObj)
      dataEs.Estimates = parsed.Estimates[d]
      dataEstList.push(dataEs.Estimates)
    }

    SpreadList = parsed.Spread.filter(function (el: null) {
      return el != null
    })

    //ticks - in Y-axis .
    var min = Math.min(...SpreadList)
    var max = Math.max(...SpreadList)
    let interval = 3
    let diff = (max - min) / interval
    var ticks1 = []
    let start = min
    ticks1.push(Math.ceil(start))
    for (let i = 0; i < interval; i++) {
      ticks1.push(Math.round(start + diff))
      start += diff
    }
    ticks1.push(0)
    ticks1.sort((a, b) => a - b)
    Math.round(ticks1[ticks1.length - 1])
    //yaxisLabel -length
    var longestLabelLength = `${max}`.length
    if (longestLabelLength < 6) {
      longestLabelLength = 6
    }
  }

  useEffect(() => {
    let titleDropdownData: titleDropdownModel[] = []
    const getIndexing = (e: object) => {
      let parsed1 = e
      setIndex(0)
      for (let [key, value] of Object.entries(parsed1)) {
        let maxDate2: number
        let minDate1: number
        if (value && value.Period) {
          minDate1 = new Date(value.Period[0]).getTime()
          var filteredPeriod = value.Period.filter(function (el: string) {
            return el != null
          })
        }
        if (
          selectValue === ChartConstant.ANNUAL_DATA ||
          selectValue === ChartConstant.CUMULATIVE_DATA
        ) {
          let date = Moment(filteredPeriod[filteredPeriod.length - 1])
            .add(30, 'day')
            .add(24, 'hours')

          maxDate2 = new Date(date).getTime()
        } else {
          maxDate2 = new Date(
            filteredPeriod[filteredPeriod.length - 1]
          ).getTime()
        }
        if (selectedUnitId === value[powerPlantUnitId]) {
          titleDropdownData.push({
            label: value[powerTitle],
            value: value[powerTitle],
          })
        }
        let d = new Date(startDate)
        d.setHours(0)
        d.setMinutes(0)
        d.setSeconds(0)
        d.setMilliseconds(0)
        let date = d.getTime()
        if (date >= minDate1 && date <= maxDate2) {
          setIndex(key)
          break
        }
      }
      setTitleDropdown([...titleDropdownData])
    }
    const getDataOfPreNext = (preNextButton: boolean) => {
      switch (selectValue) {
        case ChartConstant.ANNUAL_DATA:
          if (!preNextButton) {
            getIndexing(propsData?.Annual)
          }
          setLinechartData(
            FetchData(
              propsData?.Annual,
              indexref.current,
              startDate,
              ChartConstant.ANNUAL_DATA
            )
          )
          break
        case ChartConstant.MONTHLY_DATA:
          if (!preNextButton) {
            getIndexing(propsData?.Monthly)
          }
          setLinechartData(
            FetchData(
              propsData?.Monthly,
              indexref.current,
              startDate,
              ChartConstant.MONTHLY_DATA
            )
          )
          break

        default:
          if (!preNextButton) {
            getIndexing(propsData?.Annual)
          }
          setLinechartData(
            FetchData(
              propsData?.Annual,
              indexref.current,
              startDate,
              ChartConstant.ANNUAL_DATA
            )
          )
      }
    }
    getDataOfPreNext(false)
  }, [
    selectValue,
    startDate,
    setIndex,
    indexref,
    propsData?.Annual,
    propsData?.Monthly,
    selectedUnitId,
    i18n,
    powerPlantUnitId,
    powerTitle
  ])
  
  const handleGeneratorSelection = (e: any) => {
    setSelectedUnitId(getFormattedUnitId(e.value))
    setSelectedTitleDropdown(spreadLineChartdata[0]?.poweTitle)
  }
  const onSelectHandler = (e: { label: string; value: string }) => {
    setSelectValue(e.value)
  
  }
  const handleTitleEvent = (e: React.ChangeEvent<{ value: string }>) => {
    setSelectedTitleDropdown(e.value)
  }
  const getDataOfPreNext = () => {
    switch (selectValue) {
      case ChartConstant.ANNUAL_DATA:
        setLinechartData(
          FetchData(
            propsData?.Annual,
            indexref.current,
            startDate,
            ChartConstant.ANNUAL_DATA
          )
        )
        break

      case ChartConstant.MONTHLY_DATA:
        setLinechartData(
          FetchData(
            propsData?.Monthly,
            indexref.current,
            startDate,
            ChartConstant.MONTHLY_DATA
          )
        )
        break
      default:
        setLinechartData(
          FetchData(
            propsData?.Annual,
            indexref.current,
            startDate,
            ChartConstant.ANNUAL_DATA
          )
        )
    }
  }

  useEffect(() => {
    setSelectedTitleDropdown(titleDropdown && titleDropdown[0]?.value)
  }, [titleDropdown])

  //Prev
  let onClickPrev = () => {
    setIndex(parseInt(ind) - 1)
    getDataOfPreNext()
  }
  //Next
  let onClickNext = () => {
    setIndex(parseInt(ind) + 1)
    getDataOfPreNext()
  }

  let titleDropdownValue =titleDropdown&& titleDropdown.filter( (ele :any, ind) => ind === titleDropdown.findIndex( elem => elem.label === ele.label && elem.value === ele.value))

  return (
    <div>
      <h5 className="mb-24 mt-24 font-18 uk-text-medium">
        {i18n.t('UNIT_OPERATION.SPECIFICATION_SPREAD')}
      </h5>
      <div className="uk-flex uk-flex-middle uk-flex-wrap mb-20 align-end">
        <div className="uk-flex uk-flex-center">
          {ApiParamUtils.unitName() === AppConstant.OVERVIEW ? (
            <GeneratorDropDown
              onSelectHandler={handleGeneratorSelection}
              selectValue={selectedUnitId}
              data={generatorDropdownUnit()}
              label={i18n.t(
                'UNIT_OPERATION.DURATION_OF_SPREAD_TABLE_HEADER.GENERATOR'
              )}
              customClass="width-100"
            />
          ) : null}
         
          <GeneratorDropDown
            onSelectHandler={handleTitleEvent}
            selectValue={selectedTitleDropdown}
            data= {titleDropdownValue}
            label={i18n.t('UNIT_OPERATION.DROPDOWN_TITLE')}
            customClass="width-100"
          /> 
          <LinechartDropDown
            onSelectHandler={onSelectHandler}
            selectValue={selectValue}
            cumulativeDisable={false}
            customClass="align-end mb-0"
            isSpreadOptions={true}
          />

        </div>
        <DataChangeOptions
          linechartData={linechartData}
          onClickPrev={onClickPrev}
          startDate={startDate}
          selectValue={selectValue}
          setStartDate={setStartDate}
          onClickNext={onClickNext}
          ind={ind}
          customClass={`${
            linechartData?.currentDisable === true
              ? 'date-picker-specification'
              : 'date-picker-return'
          }`}
        />
      </div>
      <SpreadLineChart
        data={spreadLineChartdata}
        selectValue={selectValue}
        list={list}
        selectedUnitId={selectedUnitId}
        selectedTitleDropdown={selectedTitleDropdown}
      />
    </div>
  )
}

export default SpecificationSpread
