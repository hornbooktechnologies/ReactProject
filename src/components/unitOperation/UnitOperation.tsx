import React, { useEffect, useState } from "react";
import { useTranslation } from 'react-i18next';
import { Kpi_005ApiRes, UnitOperationBaseModel } from "../../models/unitOperation/UnitOperationModel";
import { unitOperationApiMockData, unitOperationMockData } from "../../models/unitOperation/_mockData";
import LoadingError from "../common/LoadingError";
import PageHeader from "../common/PageHeader";
import Spinner from "../common/Spinner";
import AmountOfOperatingTime from "./amountOfOperatingTime/AmountOfOperatingTime";
import AverageSpread from "./averageSpread/AverageSpread";
import SpreadNumberOfCases from "./spreadNumberOfCases/SpreadNumberOfCases";
import { spreadOperationMockData } from "../../models/unitOperation/_mockData";
import NegativeOperationGrossMargin from "./NegativeOperationGrossMargin/NegativeGrossMarginTable";
import ReferenceSpecificationOfSpreadOperation from "./specificationSpread/ReferenceSpecificationOfSpreadOperation";
import SpecificationSpread from "./specificationSpread/SpecificationSpread";
import Top20DurationOfSpread from "./top20DurationOfSpread/Top20DurationOfSpread";
import { getCutoffFractionTime } from "../../utils/utils";
import { negOperationAvrgSpreadApiResponse, negOperationGrossmarginApiResponse, negOperationTimeApiResponse } from "../../utils/unitOperation/UnitOperationUtils";
import { NegativeOperationTimeModel} from "../../models/unitOperation/NegOperationTimeModel";
import { NegativeOperationAvgSpreadModel } from "../../models/unitOperation/NegOperationAvgSpreadModel.ts";
import { NegativeGrossMarginTableModel } from "../../models/unitOperation/NegOperationGrossmarginModel";
import "../../assets/scss/annualTarget/_annualstatus.scss";
import '../../assets/scss/unitOperation/unitOperation.scss';
import { DpmService } from "../../services/DpmServices";
import { ApiParamUtils } from "../../utils/apiUtils/ApiParamUtils";

const UnitOperation = () => {
  const { i18n } = useTranslation();
  const [unitOperationData, setUnitOperationData] = useState({} as UnitOperationBaseModel);
  const [negOperationGrossMarginData, setNegOperationGrossMargin] = useState({} as NegativeGrossMarginTableModel)
  const [negOperationTimeData, setNegOperationTime] = useState([] as NegativeOperationTimeModel)
  const [negOperationAvrgSpreadData, setNegOperationAvrgSpread] = useState([] as NegativeOperationAvgSpreadModel)
  const [unitOperationApiData, setUnitOperationApiData] = useState({});
  const [errorThrown, seterrorThrown] = useState(null)
  const [kpiResponse, setKpiResponse] = useState(false)
  const [time, setNoDataDate] = useState(
    getCutoffFractionTime()
  )

  const noDataHandler = (val: number, selectValue: string) => {
    setNoDataDate(val)
  }
    
  useEffect(() => {
    DpmService.fetchKPI_005Data(ApiParamUtils.plantId(),ApiParamUtils.unitCode()) 
    .then((res)=>{
      res=res.data
      if (res.Success) {
        setKpiResponse(res.Success)
        let kpiResponse = res.Data as Kpi_005ApiRes;
        let mappedNegGrossMarginResponse = negOperationGrossmarginApiResponse( kpiResponse);
        let mappedNegOperationTimeResponse = negOperationTimeApiResponse(i18n, kpiResponse);
        let mappedNegAverageSpread = negOperationAvrgSpreadApiResponse(i18n , kpiResponse);
        setNegOperationGrossMargin(mappedNegGrossMarginResponse)
        setNegOperationTime(mappedNegOperationTimeResponse)
        setNegOperationAvrgSpread(mappedNegAverageSpread)
       }
    })
    .catch((error) => {
      setKpiResponse(false)
      seterrorThrown(error)
     })
    setUnitOperationData(unitOperationMockData)
    setUnitOperationApiData(unitOperationApiMockData)
  }, [i18n])

  if (kpiResponse === true && Object.entries(unitOperationData && unitOperationApiData).length > 0) { 
    return (
      <>
        <PageHeader
          title={i18n.t('UNIT_OPERATION.TITLE')}
          description={i18n.t('UNIT_OPERATION.DESCRIPTION')}
        />
        <NegativeOperationGrossMargin negativeGrossMargin = {negOperationGrossMarginData}/>
        <AmountOfOperatingTime negOperationTime = {negOperationTimeData} />
        <AverageSpread negOperationAvrgSpread = {negOperationAvrgSpreadData} />
        <SpecificationSpread spreadOperation={spreadOperationMockData} time = {time} noDataHandler={noDataHandler}/>
        <ReferenceSpecificationOfSpreadOperation {...unitOperationData.referenceSpecification} />
        <SpreadNumberOfCases spreadBarData={unitOperationData.spreadBarChart}/>
        <Top20DurationOfSpread {...unitOperationData.top20DurationOfSpread} />
      </>
    );
  } else if (errorThrown != null) {
    return <LoadingError />;
  } else {
    return <Spinner />;
  }
};
export default UnitOperation;
