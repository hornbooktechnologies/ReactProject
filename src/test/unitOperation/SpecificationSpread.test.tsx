import React from "react";
import renderer from "react-test-renderer";
import SpecificationSpread from "../../components/unitOperation/specificationSpread/SpecificationSpread";
import i18n from "../../i18nForTests";
import { I18nextProvider } from "react-i18next";
import { getCutoffFractionTime } from "../../utils/utils";
import { spreadOperationMockData } from "../../models/unitOperation/_mockData";


describe("Component: SpecificationSpread", () => {
  it("SpecificationSpread renders correctly", () => {
    const noDataHandler = (val: number, selectValue: string) => {
      }
    const tree = renderer
      .create(
        <I18nextProvider i18n={i18n}>
          <SpecificationSpread time={getCutoffFractionTime()} spreadOperation={spreadOperationMockData} noDataHandler={noDataHandler}/>
        </I18nextProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
