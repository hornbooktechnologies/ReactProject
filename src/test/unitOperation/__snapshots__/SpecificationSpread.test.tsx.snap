// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: SpecificationSpread SpecificationSpread renders correctly 1`] = `
<div>
  <h5
    className="mb-24 mt-24 font-18 uk-text-medium"
  >
    UNIT_OPERATION.SPECIFICATION_SPREAD
  </h5>
  <div
    className="uk-flex uk-flex-middle uk-flex-wrap mb-20 align-end"
  >
    <div
      className="uk-flex uk-flex-center"
    >
      <div
        className="mr-24 dropdown-select form-control"
      >
        <label
          className="uk-custom-label"
        >
          <span>
            UNIT_OPERATION.DROPDOWN_TITLE
          </span>
        </label>
        <div
          className=" css-b62m3t-container"
          onKeyDown={[Function]}
        >
          <span
            className="css-1f43avz-a11yText-A11yText"
            id="react-select-2-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            className="css-1f43avz-a11yText-A11yText"
          />
          <div
            className="form_select_input__control css-1s2u09g-control"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <div
              className="form_select_input__value-container css-319lph-ValueContainer"
            >
              <div
                className="form_select_input__placeholder css-14el2xx-placeholder"
                id="react-select-2-placeholder"
              >
                Select...
              </div>
              <input
                aria-autocomplete="list"
                aria-controls="react-select-2-listbox"
                aria-describedby="react-select-2-placeholder"
                aria-expanded={false}
                aria-haspopup={true}
                aria-owns="react-select-2-listbox"
                aria-readonly={true}
                className="css-mohuvp-dummyInput-DummyInput"
                disabled={false}
                id="react-select-2-input"
                inputMode="none"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                role="combobox"
                tabIndex={0}
                value=""
              />
            </div>
            <div
              className="form_select_input__indicators css-1hb7zxy-IndicatorsContainer"
            >
              <span
                className="form_select_input__indicator-separator css-1okebmr-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                className="form_select_input__indicator form_select_input__dropdown-indicator css-tlfecz-indicatorContainer"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="css-tj5bde-Svg"
                  focusable="false"
                  height={20}
                  viewBox="0 0 20 20"
                  width={20}
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="form-control"
      >
        <label
          className="uk-custom-label"
        >
          <span>
            LABELS.DISPLAY
          </span>
        </label>
        <div
          className="align-end mb-0 css-b62m3t-container"
          onKeyDown={[Function]}
        >
          <span
            className="css-1f43avz-a11yText-A11yText"
            id="react-select-3-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            className="css-1f43avz-a11yText-A11yText"
          />
          <div
            className="form_select_input__control css-1s2u09g-control"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <div
              className="form_select_input__value-container form_select_input__value-container--has-value css-319lph-ValueContainer"
            >
              <div
                className="form_select_input__single-value css-qc6sy-singleValue"
              >
                DROPDOWN_OPTIONS.ANNUAL_VIEW
              </div>
              <input
                aria-autocomplete="list"
                aria-controls="react-select-3-listbox"
                aria-expanded={false}
                aria-haspopup={true}
                aria-owns="react-select-3-listbox"
                aria-readonly={true}
                className="css-mohuvp-dummyInput-DummyInput"
                disabled={false}
                id="react-select-3-input"
                inputMode="none"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                role="combobox"
                tabIndex={0}
                value=""
              />
            </div>
            <div
              className="form_select_input__indicators css-1hb7zxy-IndicatorsContainer"
            >
              <span
                className="form_select_input__indicator-separator css-1okebmr-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                className="form_select_input__indicator form_select_input__dropdown-indicator css-tlfecz-indicatorContainer"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="css-tj5bde-Svg"
                  focusable="false"
                  height={20}
                  viewBox="0 0 20 20"
                  width={20}
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="uk-margin-auto-left uk-flex uk-flex-middle uk-flex-wrap uk-width-1-1 uk-width-auto@s mt-15 top-m1"
    >
      <div
        className="uk-flex uk-flex-middle uk-width-1-1 uk-width-auto@s"
      >
        <button
          className="uk-button-outline"
          onClick={[Function]}
        >
          <i
            className="jera-arrow-left-2 color-white"
          />
        </button>
        <div
          className="uk-field-with-icon ml-10 m-w-200 uk-width-1-1 uk-width-auto@s mt-m14"
        >
          <div
            className="form-control"
          >
            <label
              className="uk-custom-label"
            >
              <span>
                Year
              </span>
            </label>
          </div>
          <div
            className="cal-inline"
          >
            <p
              className="m-0 pr-4"
            >
              undefined
            </p>
          </div>
          <label
            className="mb-0"
          >
            <div
              className="react-datepicker-wrapper"
            >
              <div
                className="react-datepicker__input-container"
              >
                <input
                  className=""
                  disabled={false}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  readOnly={false}
                  type="text"
                  value="03/21/2022"
                />
              </div>
            </div>
            <i
              className="jera-calendar-1 icn top-26"
              onClick={[Function]}
            />
          </label>
        </div>
        <button
          className="uk-button-outline ml-10"
          onClick={[Function]}
        >
          <i
            className="jera-arrow-right-3 color-white"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    className="uk-card mb-24 mt-24"
    uk-card="true"
  >
    <div
      className="noDataText uk-flex uk-flex-center"
    >
      <h2
        className="mb-0"
      >
        ANNUAL_TARGET.NO_DATA_AVAILABLE
      </h2>
    </div>
  </div>
</div>
`;
